apiserver/common/storagecommon/blockdevices.go

```diff
@@ -41,77 +41,99 @@ func MatchingBlockDevice(
 	attachmentInfo state.VolumeAttachmentInfo,
 	planBlockInfo state.BlockDeviceInfo,
 ) (*state.BlockDeviceInfo, bool) {
-	logger.Tracef("looking for block device for volume %#v", volumeInfo)
-	for _, dev := range blockDevices {
-		if planBlockInfo.HardwareId != "" {
+	logger.Tracef("looking for block device to match one of planBlockInfo %#v volumeInfo %#v attachmentInfo %#v",
+		planBlockInfo, volumeInfo, attachmentInfo)
+
+	if planBlockInfo.HardwareId != "" {
+		for _, dev := range blockDevices {
 			if planBlockInfo.HardwareId == dev.HardwareId {
-				logger.Tracef("plan hwid match on %v", volumeInfo.HardwareId)
+				logger.Tracef("plan hwid match on %v", planBlockInfo.HardwareId)
 				return &dev, true
 			}
 		}
-		if planBlockInfo.WWN != "" {
+		logger.Tracef("no match for block device hardware id: %v", planBlockInfo.HardwareId)
+	}
+
+	if planBlockInfo.WWN != "" {
+		for _, dev := range blockDevices {
 			if planBlockInfo.WWN == dev.WWN {
-				logger.Tracef("plan wwn match on %v", volumeInfo.WWN)
+				logger.Tracef("plan wwn match on %v", planBlockInfo.WWN)
 				return &dev, true
 			}
-			continue
```

Comment:
It seems the crux of the change in this PR is to remove a couple of places where we continue the loop without the chance performing other matches. I am not sure about the other refactoring to introduce several places where the block devices are iterated. Won't we fix the bug simply by omitting the continues and keeping the loop logic the same?

Reply:
Yes but removing the continues affects the priority of the matching. By breaking out each match into their own loops, priority is maintained while also not skipping the lower priority match.